indico/modules/events/models/reviews.py

```diff
@@ -84,44 +94,74 @@ def get_reviewer_render_data(self, user):
                 'reviews': reviews}
 
 
+class ProposalMixin(object):
+    proposal_type = None
+    call_for_proposals_attr = None
+    revisions_enabled = True
+
+    # endpoints
+    delete_comment_endpoint = None
+    create_comment_endpoint = None
+    edit_comment_endpoint = None
+    create_review_endpoint = None
+    edit_review_endpoint = None
+
+    @property
+    def cfp(self):
+        return getattr(self.event_new, self.call_for_proposals_attr)
+
+    def get_revisions(self):
+        if self.revisions_enabled:
+            raise NotImplementedError
+        else:
+            return [self]
+
+    def get_last_revision(self):
+        if self.revisions_enabled:
+            raise NotImplementedError
+        else:
+            return self
+
+    def can_comment(self, user):
+        raise NotImplementedError
+
+    def can_review(self, user, check_state=False):
+        raise NotImplementedError
+
+    def get_delete_comment_url(self, comment):
+        return url_for(self.delete_comment_endpoint, comment)
+
+    def get_save_comment_url(self, comment=None):
+        return (url_for(self.edit_comment_endpoint, comment)
+                if comment
+                else url_for(self.create_comment_endpoint, self))
+
+    def get_save_review_url(self, group=None, review=None):
+        return (url_for(self.edit_review_endpoint, review)
+                if review
+                else url_for(self.create_review_endpoint, self, group))
+
+
 class ProposalCommentMixin(object):
     timeline_item_type = 'comment'
 
     def can_edit(self, user):
         raise NotImplementedError
 
 
-class ProposalCommentVisibility(RichIntEnum):
-    """Most to least restrictive visibility for abstract comments"""
-    __titles__ = [None,
-                  _("Visible only to judges"),
-                  _("Visible to conveners and judges"),
-                  _("Visible to reviewers, conveners, and judges"),
-                  _("Visible to contributors, reviewers, conveners, and judges"),
-                  _("Visible to all users")]
-    judges = 1
-    conveners = 2
-    reviewers = 3
-    contributors = 4
-    users = 5
-
-
 class ProposalReviewMixin(object):
     timeline_item_type = 'review'
-    proposal_relationship = None
+    revision_attr = None
     group_relationship = None
```

Comment:
I would rename `group_relationship` to `group_relationship` not to make unnecessary assumptions.